csrc/mmdeploy/core/registry.h

```diff
@@ -63,67 +106,123 @@ class MMDEPLOY_API Registry<void> {
-  template <>                                \
-  Registry<EntryType>& Registry<EntryType>::Get();
+// Specify creator signature for tag
+#define MMDEPLOY_CREATOR_SIGNATURE(tag, signature) \
+  ::mmdeploy::basic_type<signature> get_signature(::mmdeploy::basic_type<tag>);
 
-#define MMDEPLOY_DEFINE_REGISTRY(EntryType)                         \
-  template <>                                                       \
-  MMDEPLOY_EXPORT Registry<EntryType>& Registry<EntryType>::Get() { \
-    static Registry v;                                              \
-    return v;                                                       \
+#define MMDEPLOY_DECLARE_REGISTRY(tag, signature) \
+  MMDEPLOY_CREATOR_SIGNATURE(tag, signature)      \
+  MMDEPLOY_API ::mmdeploy::Registry<tag>& get_registry(::mmdeploy::basic_type<tag>);
+
+#define MMDEPLOY_DECLARE_REGISTRY_EXPAND(tag, signature)        \
+  MMDEPLOY_CREATOR_SIGNATURE(tag, MMDEPLOY_PP_EXPAND signature) \
+  MMDEPLOY_API ::mmdeploy::Registry<tag>& get_registry(::mmdeploy::basic_type<tag>);
+
+#define MMDEPLOY_DEFINE_REGISTRY(tag)                                    \
+  ::mmdeploy::Registry<tag>& get_registry(::mmdeploy::basic_type<tag>) { \
+    static ::mmdeploy::Registry<tag> instance{};                         \
+    return instance;                                                     \
   }
 
-#define REGISTER_MODULE(EntryType, CreatorType) \
-  static ::mmdeploy::Registerer<EntryType, CreatorType> g_register_##EntryType##_##CreatorType{};
-
-#define DECLARE_AND_REGISTER_MODULE(base_type, module_name, version)   \
-  class module_name##Creator : public ::mmdeploy::Creator<base_type> { \
-   public:                                                             \
-    module_name##Creator() = default;                                  \
-    ~module_name##Creator() = default;                                 \
-    const char* GetName() const override { return #module_name; }      \
-    int GetVersion() const override { return version; }                \
-                                                                       \
-    std::unique_ptr<base_type> Create(const Value& value) override {   \
-      return std::make_unique<module_name>(value);                     \
-    }                                                                  \
-  };                                                                   \
-  REGISTER_MODULE(base_type, module_name##Creator);
+#define MMDEPLOY_REGISTER_CREATOR(tag, creator_type)                    \
+  static ::mmdeploy::Registerer MMDEPLOY_ANONYMOUS_VARIABLE(register_){ \
+      [creator = creator_type{}]() mutable { ::mmdeploy::gRegistry<tag>().Add(creator); }};
+
+#define MMDEPLOY_CREATOR_DESC(name, version) #name, version
+
+#define MMDEPLOY_REGISTER_FACTORY_FUNC(tag, creator_desc, func)                                  \
+  static ::mmdeploy::Registerer MMDEPLOY_ANONYMOUS_VARIABLE(register_){                          \
```

Comment:
`register_` is a static variable. Is it necessary to make it globally identical?

Reply:
depending on whether the `__COUNTER__` macro is supported, `MMDEPLOY_ANONYMOUS_VARIABLE(register_)` is going to be expanded to something like
```
register_0
register_1
register_2
```
or
```
register_ ## __LINE__
```
so as long as you don't put multiple `MMDEPLOY_REGISTER_FACTORY_FUNC` in the same line, it will be fine.